cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,222 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+type RolloutStatus int
+
+const (
+	// resource desired status is not upgraded yet
+	ToUpgrade RolloutStatus = iota
+	// resource desired status is upgraded and last applied status is not upgraded
+	Upgrading
+	// resource desired status is upgraded and last applied status is upgraded successfully
+	UpgradeSucceed
+	// resource desired status is upgraded and last applied status upgrade failed
+	UpgradeFailed
+)
+
+// +k8s:deepcopy-gen=false
+type RolloutHandler struct {
+	// duck type rollout strategy
+	rolloutStrategy RolloutStrategy
+	// placement decision tracker
+	pdTpdTracker clusterv1beta1.PlacementDecisionClustersTracker
+	// Record the resource rollout status on each managed cluster
+	// For example, map[string]RolloutStatus{"cluster1":UpgradeSucceed, "cluster2":Upgrading, "cluster3":ToUpgrade}
+	objs map[string]RolloutStatus
+}
+
+func NewRolloutHandler(rolloutStrategy RolloutStrategy, pdTpdTracker clusterv1beta1.PlacementDecisionClustersTracker, objs map[string]RolloutStatus) *RolloutHandler {
+	return &RolloutHandler{
+		rolloutStrategy: rolloutStrategy,
+		pdTpdTracker:    pdTpdTracker,
+		objs:            objs,
+	}
+}
+
+func (r *RolloutHandler) GetRolloutCluster() []string {
```

Comment:
it is not quite clear on what is the output, is this the list cluster that needs to be updated with current state? If so I think the the input should have
1. the strategy
2. the current state.

Reply:
Update the output.